### PR TITLE
Custom parameters support + Hook conflict fix

### PR DESCRIPTION
**Pull request by BugDiver**
* Add support for custom parameter parsing #5
* Fix conflicting hooks issues #138

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -309,6 +309,63 @@ String elementId = scenarioStore.get("element-id") as string;
 
 ```
 
+### Custom Paramter Parsers
```
Comment (zabil) [inline docs/index.md:312]:
```suggestion
### Custom Parameter Parsers
```

File: docs/index.md [hunk 2]
```diff
@@ -309,6 +309,63 @@ String elementId = scenarioStore.get("element-id") as string;
...
+
+* By default gauge-ts tries to convert the spec parameter to primitives (number, boolean), for table parameters gauge-ts will convert paramters
+to `Table` type. gauge-ts also provide a way to customize the paramter parsing.
+If you need to have custom paramters in your step implementations create a paramter parser which should implements methods from `ParamterParser` interface.
```
Comment (zabil) [inline docs/index.md:316]: Should this be 
```suggestion
If you need to have custom paramters in your step implementations create a parameter parser which should implements methods from `ParameterParser` interface.
```

?

File: docs/index.md [hunk 3]
```diff
@@ -309,6 +309,63 @@ String elementId = scenarioStore.get("element-id") as string;
...
+
+export default class Implementation {
+    @Step
+    public async isAdult(person: Person)y {
+      assert.ok(person.isAdult())
+    }
+}
+
+```
+Cusotm ParameterParser:
```
Comment (zabil) [inline docs/index.md:350]:
```suggestion
Custom ParameterParser:
```

File: e2e/tests/PersonParameterParser.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { Person } from "@lib/Person";
+import type { Parameter, ParameterParser } from "gauge-ts";
+export default class PersonParameterParser implements ParameterParser {
+  public canParse(paramter: Parameter): boolean {
+    return (
+      paramter.getValue().startsWith("{") && paramter.getValue().endsWith("}")
```
Comment (zabil) [inline e2e/tests/PersonParameterParser.ts:6]:
```suggestion
      parameter.getValue().startsWith("{") && parameter.getValue().endsWith("}")
```

File: gauge-ts/src/RunnerServer.ts [hunk 5]
```diff
@@ -141,7 +146,7 @@ export default class RunnerServer implements IRunnerServer {
   ): void {
     try {
       DataStoreFactory.getSuiteDataStore().clear();
-      const loader = new ImplLoader();
+      const loader = new ImplLoader(RunnerServer.paramterParsingChain);
```
Comment (zabil) [inline gauge-ts/src/RunnerServer.ts:149]:
```suggestion
      const loader = new ImplLoader(RunnerServer.parameterParsingChain);
```

File: gauge-ts/src/RunnerServer.ts [hunk 6]
```diff
@@ -109,7 +112,9 @@ export default class RunnerServer implements IRunnerServer {
       new SpecExecutionStartingProcessor();
     RunnerServer.stepExecutionEndingProcessor =
       new StepExecutionEndingProcessor();
-    RunnerServer.stepExecutionProcessor = new StepExecutionProcessor();
+    RunnerServer.stepExecutionProcessor = new StepExecutionProcessor(
+      RunnerServer.paramterParsingChain,
```
Comment (zabil) [inline gauge-ts/src/RunnerServer.ts:116]:
```suggestion
      RunnerServer.parameterParsingChain,
```

File: gauge-ts/src/loaders/ImplLoader.ts [hunk 7]
```diff
@@ -9,18 +11,28 @@ type ModuleType = {
...
       try {
         process.env.STEP_FILE_PATH = file;
         const c = (await Util.importFile(file)) as ModuleType;
-
         if (c.default && c.default.length === 0) {
           const instance = new c.default();
-
-          ImplLoader.updateRegistry(file, instance);
+          if (Util.isCustomParameterParser(instance)) {
+            this.paramterParsignChain.addCustomParser(
```
Comment (zabil) [inline gauge-ts/src/loaders/ImplLoader.ts:30]:
```suggestion
            this.parameterParsingChain.addCustomParser(
```

File: docs/index.md [hunk 8]
```diff
@@ -309,6 +309,63 @@ String elementId = scenarioStore.get("element-id") as string;
 
 ```
 
+### Custom Paramter Parsers
+
+* By default gauge-ts tries to convert the spec parameter to primitives (number, boolean), for table parameters gauge-ts will convert paramters
```
Comment (zabil) [inline docs/index.md:314]:
```suggestion
* By default gauge-ts tries to convert the spec parameter to primitives (number, boolean), for table parameters gauge-ts will convert parameters
```

File: docs/index.md [hunk 9]
```diff
@@ -309,6 +309,63 @@ String elementId = scenarioStore.get("element-id") as string;
...
+to `Table` type. gauge-ts also provide a way to customize the paramter parsing.
```
Comment (zabil) [inline docs/index.md:315]:
```suggestion
to `Table` type. gauge-ts also provide a way to customize the parameter parsing.
```

**Review by zabil — CHANGES_REQUESTED** [review]
Taking a look at this meanwhile I noticed what might be typos. Can you check if it's worth changing the spelling?

File: gauge-ts/tests/loaders/ImplLoaderTests.ts [hunk 10]
```diff
@@ -44,5 +52,17 @@ describe("ImplLoader", () => {
       await loader.loadImplementations();
       expect(exception).toContain("failed to import");
     });
+
+    it("shold load custom parameter parser", async () => {
```
Comment (zabil) [inline gauge-ts/tests/loaders/ImplLoaderTests.ts:56]:
```suggestion
    it("should load custom parameter parser", async () => {
```

File: e2e/tests/implementation.ts [hunk 11]
```diff
@@ -31,4 +33,10 @@ class Implementation {
       assert.equal(Implementation.vowelsCount(word), Number.parseInt(count));
     }
   }
+  @Step("This step uses a custom parameter of type Person and value <person>")
+  public async validatePerson(person: Person) {
+    assert.equal(person.name, "John");
+    assert.equal(person.age, 30);
```
Comment (zabil) [inline e2e/tests/implementation.ts:39]: Can we change these to use `strictEqual`? I noticed that even if the parameter was declared as a `number` in earlier versions the number is actually passed as `string` and fails when we do a `===`

**Review by zabil — CHANGES_REQUESTED** [review]
Minor comments otherwise all else looks ok!

File: e2e/tests/implementation.ts (hunk 11 above)
Comment (BugDiver) [reply e2e/tests/implementation.ts]: Yes, this should be fix now, as we are doing primitive conversion. Just to highlight, parameters which are part of a table (string or int or bool) are still treated as `string`. As of now we are doing nested conversion.

File: gauge-ts/src/loaders/ImplLoader.ts [hunk 12]
```diff
@@ -11,10 +11,10 @@ type ModuleType = {
 };
 
 export class ImplLoader {
-  private paramterParsignChain: ParameterParsingChain;
+  private parameterParsignChain: ParameterParsingChain;
```
Comment (zabil) [inline gauge-ts/src/loaders/ImplLoader.ts:14]: Sorry to be irritating. But should this be
```suggestion
  private parameterParsingChain: ParameterParsingChain;
```

File: gauge-ts/src/loaders/ImplLoader.ts [hunk 13]
```diff
@@ -27,7 +27,7 @@ export class ImplLoader {
         if (c.default && c.default.length === 0) {
           const instance = new c.default();
           if (Util.isCustomParameterParser(instance)) {
-            this.paramterParsignChain.addCustomParser(
+            this.parameterParsignChain.addCustomParser(
```
Comment (zabil) [inline gauge-ts/src/loaders/ImplLoader.ts:30]:
```suggestion
            this.parameterParsingChain.addCustomParser(
```

**Review by zabil — CHANGES_REQUESTED** [review]
Two more typos otherwise all good.

**Review by zabil — APPROVED** [review]

**Pull request merged**